flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java

```diff
@@ -525,56 +545,102 @@ public static Integer toInt(BinaryStringData str) {
         while (i < sizeInBytes) {
             byte currentByte = tmpBytes[i];
             if (currentByte < '0' || currentByte > '9') {
-                return null;
+                throw numberFormatExceptionFor(str, "INT", "Invalid character found");
             }
             i++;
         }
 
         if (!negative) {
             result = -result;
             if (result < 0) {
-                return null;
+                throw numberFormatExceptionFor(str, "INT", "Overflow");
             }
         }
         return result;
     }
 
-    public static Short toShort(BinaryStringData str) {
-        Integer intValue = toInt(str);
-        if (intValue != null) {
-            short result = intValue.shortValue();
-            if (result == intValue) {
-                return result;
-            }
+    public static short toShort(BinaryStringData str) throws TableException {
+        int intValue = toInt(str);
+        short result = (short) intValue;
+        if (result == intValue) {
+            return result;
         }
-        return null;
+        throw numberFormatExceptionFor(str, "SHORT", "Overflow");
     }
 
-    public static Byte toByte(BinaryStringData str) {
-        Integer intValue = toInt(str);
-        if (intValue != null) {
-            byte result = intValue.byteValue();
-            if (result == intValue) {
-                return result;
-            }
+    public static byte toByte(BinaryStringData str) throws TableException {
+        int intValue = toInt(str);
+        byte result = (byte) intValue;
+        if (result == intValue) {
+            return result;
         }
-        return null;
+        throw numberFormatExceptionFor(str, "BYTE", "Overflow");
     }
 
-    public static Double toDouble(BinaryStringData str) {
+    public static double toDouble(BinaryStringData str) throws TableException {
         try {
-            return Double.valueOf(str.toString());
+            return Double.parseDouble(str.toString());
         } catch (NumberFormatException e) {
-            return null;
+            throw new TableException("Cannot parse STRING to DOUBLE", e);
         }
     }
 
-    public static Float toFloat(BinaryStringData str) {
+    public static float toFloat(BinaryStringData str) throws TableException {
         try {
-            return Float.valueOf(str.toString());
+            return Float.parseFloat(str.toString());
         } catch (NumberFormatException e) {
-            return null;
+            throw new TableException("Cannot parse STRING to FLOAT", e);
```

Comment:
Probably it makes sense to specify what string cannot be parsed like it is done for `toBoolean`.

Reply:
Same as https://github.com/apache/flink/pull/17800#discussion_r750220243